### ISSUE TITLE
Improve direct solver performance for GPGPU backends

### ISSUE THREAD

**Issue by ddemidov** [ddemidov/amgcl#24]
See discussion with @jdumas in #23.

**Comment by ddemidov** [comment]
@jdumas, what is the exact GPGPU backend you use?

**Comment by jdumas** [comment]
I'm using `amgcl::backend::vexcl<double>`.

**Comment by ddemidov** [comment]
cfb35d4 is an attempt to solve the coarse problem on the host with SkylineLU. How does that perform for you? Here it looses slightly to the previous implementation with default parameters:


before, default parameters:
```
OCL_DEVICE=K40 ./runtime_vexcl_v1 -n 80
1. Tesla K40c (NVIDIA CUDA)

Number of levels:    4
Operator complexity: 1.56
Grid complexity:     1.13

level     unknowns       nonzeros
---------------------------------
    0       512000        3545600 (64.02%)
    1        64560        1903196 (34.37%)
    2         1610          88526 ( 1.60%)
    3           25            531 ( 0.01%)

Iterations: 14
Error:      9.82844e-09


[Profile:                    1.041 s] (100.00%)
[ self:                      0.155 s] ( 14.92%)
[  assemble:                 0.023 s] (  2.23%)
[  setup:                    0.681 s] ( 65.38%)
[   self:                    0.039 s] (  3.73%)
[    coarse operator:        0.192 s] ( 18.46%)
[    coarsest level:         0.000 s] (  0.01%)
[    move to backend:        0.305 s] ( 29.27%)
[    transfer operators:     0.145 s] ( 13.91%)
[     self:                  0.047 s] (  4.56%)
[      aggregates:           0.041 s] (  3.96%)
[      interpolation:        0.056 s] (  5.39%)
[       self:                0.049 s] (  4.70%)
[        tentative:          0.007 s] (  0.69%)
[  solve:                    0.182 s] ( 17.48%)
[   self:                    0.141 s] ( 13.50%)
[    coarse:                 0.001 s] (  0.08%)
[    prolongate:             0.014 s] (  1.35%)
[    relax:                  0.015 s] (  1.40%)
[    residual:               0.002 s] (  0.15%)
[    restrict:               0.010 s] (  0.99%)
```

before, set `coarse_enough`:
```
$ cat params.json 
{
    "precond" : {
        "coarse_enough" : 5000
    }
}
$ OCL_DEVICE=K40 ./runtime_vexcl_v1 -n 80 -p params.json 
1. Tesla K40c (NVIDIA CUDA)

Number of levels:    3
Operator complexity: 1.56
Grid complexity:     1.13

level     unknowns       nonzeros
---------------------------------
    0       512000        3545600 (64.03%)
    1        64560        1903196 (34.37%)
    2         1610          88526 ( 1.60%)

Iterations: 10
Error:      3.97572e-09


[Profile:                   14.367 s] (100.00%)
[ self:                      0.160 s] (  1.11%)
[  assemble:                 0.025 s] (  0.18%)
[  setup:                   14.023 s] ( 97.60%)
[   self:                    0.040 s] (  0.28%)
[    coarse operator:        0.191 s] (  1.33%)
[    coarsest level:        13.344 s] ( 92.87%)
[    move to backend:        0.291 s] (  2.03%)
[    transfer operators:     0.157 s] (  1.09%)
[     self:                  0.055 s] (  0.38%)
[      aggregates:           0.041 s] (  0.29%)
[      interpolation:        0.060 s] (  0.42%)
[       self:                0.053 s] (  0.37%)
[        tentative:          0.007 s] (  0.05%)
[  solve:                    0.160 s] (  1.11%)
[   self:                    0.116 s] (  0.81%)
[    coarse:                 0.010 s] (  0.07%)
[    prolongate:             0.014 s] (  0.10%)
[    relax:                  0.011 s] (  0.07%)
[    residual:               0.001 s] (  0.00%)
[    restrict:               0.009 s] (  0.06%)
```

after:
```
OCL_DEVICE=K40 ./runtime_vexcl -n 80
1. Tesla K40c (NVIDIA CUDA)

Number of levels:    3
Operator complexity: 1.56
Grid complexity:     1.13

level     unknowns       nonzeros
---------------------------------
    0       512000        3545600 (64.03%)
    1        64560        1903196 (34.37%)
    2         1610          88526 ( 1.60%)

Iterations: 10
Error:      3.97572e-09


[Profile:                    1.130 s] (100.00%)
[ self:                      0.165 s] ( 14.59%)
[  assemble:                 0.023 s] (  2.07%)
[  setup:                    0.800 s] ( 70.79%)
[   self:                    0.038 s] (  3.37%)
[    coarse operator:        0.189 s] ( 16.76%)
[    coarsest level:         0.122 s] ( 10.81%)
[    move to backend:        0.301 s] ( 26.61%)
[    transfer operators:     0.149 s] ( 13.23%)
[     self:                  0.052 s] (  4.61%)
[      aggregates:           0.039 s] (  3.43%)
[      interpolation:        0.059 s] (  5.20%)
[       self:                0.052 s] (  4.59%)
[        tentative:          0.007 s] (  0.62%)
[  solve:                    0.142 s] ( 12.56%)
[   self:                    0.053 s] (  4.67%)
[    coarse:                 0.057 s] (  5.09%)
[    prolongate:             0.011 s] (  0.98%)
[    relax:                  0.011 s] (  0.97%)
[    residual:               0.001 s] (  0.06%)
[    restrict:               0.009 s] (  0.79%)
```

**Comment by jdumas** [comment]
It seems to be working nicely! It's also slower on my machine, on this example:

Before:
```
./runtime_vexcl_v1 -n 120
1. GeForce GTX TITAN (NVIDIA CUDA)

Number of levels:    4
Operator complexity: 1.56
Grid complexity:     1.12

level     unknowns       nonzeros
---------------------------------
    0      1728000       12009600 (63.99%)
    1       206075        6464415 (34.44%)
    2         4507         293143 ( 1.56%)
    3           53           1743 ( 0.01%)

Iterations: 19
Error:      8.88777e-09


[Profile:                    2.484 s] (100.00%)
[ self:                      0.066 s] (  2.65%)
[  assemble:                 0.045 s] (  1.82%)
[  setup:                    1.817 s] ( 73.17%)
[   self:                    0.072 s] (  2.91%)
[    coarse operator:        0.919 s] ( 37.01%)
[    coarsest level:         0.000 s] (  0.01%)
[    move to backend:        0.542 s] ( 21.83%)
[    transfer operators:     0.283 s] ( 11.41%)
[     self:                  0.102 s] (  4.11%)
[      aggregates:           0.081 s] (  3.26%)
[      interpolation:        0.100 s] (  4.04%)
[       self:                0.090 s] (  3.62%)
[        tentative:          0.010 s] (  0.42%)
[  solve:                    0.555 s] ( 22.36%)
[   self:                    0.481 s] ( 19.35%)
[    coarse:                 0.001 s] (  0.04%)
[    prolongate:             0.025 s] (  1.02%)
[    relax:                  0.026 s] (  1.04%)
[    residual:               0.001 s] (  0.04%)
[    restrict:               0.022 s] (  0.87%)
```

After:
```
./runtime_vexcl -n 120
1. GeForce GTX TITAN (NVIDIA CUDA)

Number of levels:    3
Operator complexity: 1.56
Grid complexity:     1.12

level     unknowns       nonzeros
---------------------------------
    0      1728000       12009600 (63.99%)
    1       206075        6464415 (34.45%)
    2         4507         293143 ( 1.56%)

Iterations: 11
Error:      8.79808e-09


[Profile:                    3.286 s] (100.00%)
[ self:                      0.071 s] (  2.17%)
[  assemble:                 0.045 s] (  1.37%)
[  setup:                    2.766 s] ( 84.17%)
[   self:                    0.070 s] (  2.14%)
[    coarse operator:        0.898 s] ( 27.32%)
[    coarsest level:         0.982 s] ( 29.89%)
[    move to backend:        0.531 s] ( 16.15%)
[    transfer operators:     0.285 s] (  8.67%)
[     self:                  0.101 s] (  3.07%)
[      aggregates:           0.083 s] (  2.51%)
[      interpolation:        0.101 s] (  3.08%)
[       self:                0.091 s] (  2.76%)
[        tentative:          0.011 s] (  0.32%)
[  solve:                    0.404 s] ( 12.29%)
[   self:                    0.141 s] (  4.31%)
[    coarse:                 0.194 s] (  5.89%)
[    prolongate:             0.025 s] (  0.75%)
[    relax:                  0.023 s] (  0.70%)
[    residual:               0.000 s] (  0.01%)
[    restrict:               0.021 s] (  0.63%)
```


However, it also needs fewer iterations. And on the linear elasticity problem I was experimenting earlier, it's a bit faster. Well, not by much, the difference is maybe 0.5s on a total time of 14s. Could be worth providing the option though, as it all very much depends on the problem and choice of preconditionner

**Comment by ddemidov** [comment]
> Could be worth providing the option though, as it all very much depends on the problem and choice of preconditionner

That is my thought exactly. I'll add a second template parameter with direct solver class (with a default choice) to the vexcl backend (and may be to the backends).

By the way, setting `coarse_enough` parameter to 300 with the new version make it even with the old:

```
$ cat params.json 
{
    "precond" : {
        "coarse_enough" : 300
    }
}
$ OCL_DEVICE=K40 ./runtime_vexcl -n 80 -p params.json 
No protocol specified
1. Tesla K40c (NVIDIA CUDA)

Number of levels:    4
Operator complexity: 1.56
Grid complexity:     1.13

level     unknowns       nonzeros
---------------------------------
    0       512000        3545600 (64.02%)
    1        64560        1903196 (34.37%)
    2         1610          88526 ( 1.60%)
    3           25            531 ( 0.01%)

Iterations: 14
Error:      9.82844e-09


[Profile:                    0.997 s] (100.00%)
[ self:                      0.158 s] ( 15.83%)
[  assemble:                 0.024 s] (  2.38%)
[  setup:                    0.631 s] ( 63.31%)
[   self:                    0.036 s] (  3.64%)
[    coarse operator:        0.162 s] ( 16.21%)
[    coarsest level:         0.000 s] (  0.00%)
[    move to backend:        0.298 s] ( 29.86%)
[    transfer operators:     0.136 s] ( 13.60%)
[     self:                  0.048 s] (  4.78%)
[      aggregates:           0.037 s] (  3.75%)
[      interpolation:        0.050 s] (  5.07%)
[       self:                0.044 s] (  4.38%)
[        tentative:          0.007 s] (  0.69%)
[  solve:                    0.184 s] ( 18.48%)
[   self:                    0.069 s] (  6.91%)
[    coarse:                 0.076 s] (  7.63%)
[    prolongate:             0.013 s] (  1.31%)
[    relax:                  0.014 s] (  1.44%)
[    residual:               0.001 s] (  0.15%)
[    restrict:               0.010 s] (  1.04%)
```

**Comment by ddemidov** [comment]
> And on the linear elasticity problem I was experimenting earlier, it's a bit faster

Offtopic: do you use `precond.coarsening.aggr.block_size = 3` with the linear elasticity problem (assuming that the coarsener is some aggregation, the problem is 3D, and the equations are ordered by the grid points, so that the system matrix is blockwise with blocks 3x3)?

**Comment by jdumas** [comment]
I'm using my own geometric-based aggregation (hence my previous question), but I'll check the built-ins with `precond.coarsening.aggr.block_size = 3`.

**Comment by ddemidov** [comment]
Fixed by f9082a5, 7819361, and ef00419
